xc/device.ml

```diff
@@ -1430,6 +1431,107 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
 
 module Dm = struct
 
+  module QMP_Event = struct
+
+    let (pipe_r, pipe_w) = Unix.pipe ()
+
+    module Lookup = struct
+      let ftod, dtoc = Hashtbl.create 16, Hashtbl.create 16
+      let add c domid =
+        Hashtbl.replace ftod (Qmp_protocol.to_fd c) domid;
+        Hashtbl.replace dtoc domid c
+      let remove c domid =
+        Hashtbl.remove ftod (Qmp_protocol.to_fd c);
+        Hashtbl.remove dtoc domid
+      let domid_of fd = try Some (Hashtbl.find ftod fd) with Not_found -> None
+      let channel_of domid = try Some (Hashtbl.find dtoc domid) with Not_found -> None
+    end
+
+    module Epoll = Core.Linux_ext.Epoll
+    module Flags = Core.Linux_ext.Epoll.Flags
+
+    module Monitor = struct
+      let create () = (Core.Std.Or_error.ok_exn Epoll.create) ~num_file_descrs:1001 ~max_ready_events:1001
+      let add m fd = Epoll.set m fd Flags.in_
+      let remove m fd = Epoll.remove m fd
+      let wait m = Epoll.wait m ~timeout:`Never
+      let with_event ~ok = function | `Ok -> ok() | `Timeout -> debug "Shouldn't receive epoll timeout event in qmp_event_thread"
+    end
+    let m = Monitor.create ()
+
+    let monitor_path domid = Printf.sprintf "/var/run/xen/qmp-event-%d" domid
+    let debug_exn msg e = debug "%s: %s" msg (Printexc.to_string e)
+
+    let remove domid =
+      try
+        match Lookup.channel_of domid with
+        | None -> ()
+        | Some c -> Lookup.remove c domid;
```

Comment:
Perhaps `Lookup.channel_of domid >>= fun c ->` , and you move >>= to
```
module QMP_Event = struct
  let (>>=) m f = match m with | Some x -> f x | None -> () in
```

Reply:
Done, thanks.